### PR TITLE
fix: always inline and deduplicate fragment definitions

### PR DESCRIPTION
**Pull request by n1ru4l**
The option `dedupeFragments` has been problematic from the beginning and caused a lot of unexpected behavior that lead to either duplicated fragment definitions within a document node or missing fragment definitions within a document node.

***ATTENTION !!!*** This PR basically replaces the existing definitions spread (`...`) with a duplication/inline strategy. While this increases the bundle size of the client applications, it is actually the only possible way of solving this issue once and for all. 

I marked this change as a fix, though it might be debatable that this should instead be a breaking change, which we could also ship IMHO.

```diff
-    ...VoteButtonsFragmentDoc.definitions,
-    ...RepoInfoFragmentDoc.definitions,
+   {
+     kind: 'FragmentDefinition',
+     name: { kind: 'Name', value: 'VoteButtons' },
+     typeCondition: { kind: 'NamedType', name: { kind: 'Name', value: 'Entry' } },
+     directives: [],
+     selectionSet: {
+       kind: 'SelectionSet',
+       selections: [
+         { kind: 'Field', name: { kind: 'Name', value: 'score' }, arguments: [], directives: [] },
+         {
+           kind: 'Field',
+           name: { kind: 'Name', value: 'vote' },
+           arguments: [],
+           directives: [],
+           selectionSet: {
+             kind: 'SelectionSet',
+             selections: [
+               { kind: 'Field', name: { kind: 'Name', value: 'vote_value' }, arguments: [], directives: [] },
+             ],
+           },
+         },
+       ],
+     },
+   },
+   {
+     kind: 'FragmentDefinition',
+     name: { kind: 'Name', value: 'RepoInfo' },
+     typeCondition: { kind: 'NamedType', name: { kind: 'Name', value: 'Entry' } },
+     directives: [],
+     selectionSet: {
+       kind: 'SelectionSet',
+       selections: [
+         { kind: 'Field', name: { kind: 'Name', value: 'createdAt' }, arguments: [], directives: [] },
+         {
+           kind: 'Field',
+           name: { kind: 'Name', value: 'repository' },
+           arguments: [],
+           directives: [],
+           selectionSet: {
+             kind: 'SelectionSet',
+             selections: [
+               { kind: 'Field', name: { kind: 'Name', value: 'description' }, arguments: [], directives: [] },
+               { kind: 'Field', name: { kind: 'Name', value: 'stargazers_count' }, arguments: [], directives: [] },
+               { kind: 'Field', name: { kind: 'Name', value: 'open_issues_count' }, arguments: [], directives: [] },
+             ],
+           },
+         },
+         {
+           kind: 'Field',
+           name: { kind: 'Name', value: 'postedBy' },
+           arguments: [],
+           directives: [],
+           selectionSet: {
+             kind: 'SelectionSet',
+             selections: [
+               { kind: 'Field', name: { kind: 'Name', value: 'html_url' }, arguments: [], directives: [] },
+               { kind: 'Field', name: { kind: 'Name', value: 'login' }, arguments: [], directives: [] },
+             ],
+           },
+         },
+       ],
+     },
+   },
```

___

Closes https://github.com/dotansimha/graphql-code-generator/issues/8967
Closes https://github.com/dotansimha/graphql-code-generator/issues/8842

### CODE REVIEW

File: packages/plugins/other/visitor-plugin-common/src/client-side-base-visitor.ts [hunk 1]
```diff
@@ -399,47 +375,50 @@ export class ClientSideBaseVisitor<
...
+      );
 
-        return `{${hashPropertyStr}"kind":"${Kind.DOCUMENT}", "definitions":[${definitions}]}`;
+      for (const fragmentName of fragmentNames) {
+        fragmentDependencyNames.add(fragmentName);
       }
 
-      let meta: ExecutableDocumentNodeMeta | void;
+      const jsonStringify = (json: unknown) =>
+        JSON.stringify(json, (key, value) => (key === 'loc' ? undefined : value));
```
Comment (n1ru4l) [inline packages/plugins/other/visitor-plugin-common/src/client-side-base-visitor.ts:397]: loc is only bloat

File: packages/presets/client/tests/client-preset.spec.ts [hunk 2]
```diff
@@ -1304,7 +1304,7 @@ export * from "./gql.js";`);
 
         export type CFragment = { __typename?: 'Query', c?: string | null } & { ' $fragmentName'?: 'CFragment' };
 
-        export const CFragmentDoc = {} as unknown as DocumentNode<CFragment, unknown>;
+        export const CFragmentDoc = {"kind":"Document","definitions":[{"kind":"FragmentDefinition","name":{"kind":"Name","value":"C"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"Query"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"c"}}]}}]} as unknown as DocumentNode<CFragment, unknown>;
```
Comment (n1ru4l) [inline packages/presets/client/tests/client-preset.spec.ts:1307]: Now this can actually be used with readFragment and stuff 😬

Comment (dotansimha) [reply packages/presets/client/tests/client-preset.spec.ts:1307]: amazing!

File: packages/plugins/other/visitor-plugin-common/src/client-side-base-visitor.ts (hunk 1 above)
Comment (dotansimha) [reply packages/plugins/other/visitor-plugin-common/src/client-side-base-visitor.ts:392]: nice

**Review by dotansimha — APPROVED** [review]

**Review by saihaj — APPROVED** [review]
Let's mark `dedupeFragments` as a deprecated option so we can drop it in the next major.

File: .changeset/rotten-lizards-bathe.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@graphql-codegen/client-preset': patch
```
Comment (n1ru4l) [inline .changeset/rotten-lizards-bathe.md:2]:
```suggestion
'@graphql-codegen/client-preset': patch
'@graphql-codegen/typed-document-node': patch
```

**Comment by JanStevens** [comment]
This is a breaking change IMO and should only be done in a major version.

Our generated file size just jumped from 600kb to a whopping 3.7MB causing issues on our frontend since there is a lot of fragment reuse going on (150 queries, lots and lots of fragments). Our Lighthouse score totally got trashed since we shipped a gigantic file, please reconsider 🙏  or keep the option in there like it was before.

We are using this in combination of typed-document-node, to be honest we never had issues with duplicate fragments (we are not using client-present atm).

Regards,

**Comment by n1ru4l** [comment]
@JanStevens Please use the SWC or babel plugin. Also, you can use persisted documents and completely drop the AST/string being shipped to the client. https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#persisted-documents

**Comment by JanStevens** [comment]
> @JanStevens Please use the SWC or babel plugin. Also, you can use persisted documents and completely drop the AST/string being shipped to the client. https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#persisted-documents

I know about persisted documents but this is not implemented everywhere. For example when querying headless CMS like Contentful or DatoCMS, they don't offer a persisted query implementation. The only option is to add your own graphql server on top of their existing implementation, which is not really maintainable.

**Comment by nandorojo** [comment]
I upgraded all my graphql-codegen versions, and I added `dedupeFragments: true` to my `client` preset. However, I still am getting an error from my server saying we're using duplicate fragments with the same name. Is there any other step I should take?

This is the error I see:

```tsx
There can be only one fragment named `MyFragment`
```


It happens when I do this:

```graphql
query Test {
  test {
    ...MyFragment
    ...Nested
  }
}

fragment Nested on Test { 
  someField
  ...MyFragment 
}
```

Thanks!

**Comment by negezor** [comment]
Unfortunately, I noticed this Pull Request too late, since `@graphql-codegen/typescript-vue-urql` had long depended on `@graphql-codegen/visitor-plugin-common` version ^2. After upgrading and regenerating types, my project size increased by 3MB, because fragment documents are no longer reused. My project has 148 queries, 216 mutations, and 203 fragments that are reused across each other.

Unfortunately, I cannot switch to `documentMode.graphQLTag` generation, as it significantly impacts TBT in Lighthouse. Fortunately, I can still stay on the latest version of graphql-codegen, but with the following override in `package.json`:

```json
{
    "overrides": {
        "@graphql-codegen/typescript-vue-urql": {
            "@graphql-codegen/visitor-plugin-common": "=3.0.0"
        }
    }
}
```

Perhaps it would be worth reconsidering this decision and bringing the option back?

**Pull request merged**